### ISSUE TITLE
[unused_declaration] False positive when the conformance to a protocol happens on an extension in a different file

### ISSUE THREAD

**Issue by acecilia** [realm/SwiftLint#3763]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

When running the `unused_declaration` analyze rule:

* Having a protocol and class conforming to that protocol : https://github.com/acecilia/swift-compilation-database/blob/master/Sources/ModuleB/SomethingProtocol.swift
* Having the implementation of the conformance in a different file: https://github.com/acecilia/swift-compilation-database/blob/master/Sources/ModuleB/Something%2Bconformance.swift

The rule will throw a false positive, detecting the conformance that happens in a different file as unused:
```
swift-compilation-database/Sources/ModuleB/Something+conformance.swift:2:10: error: Unused Declaration Violation: Declarations should be referenced at least once within all files linted. (unused_declaration)
```

### Environment

* SwiftLint version (run `swiftlint version` to be sure)? `0.45.0`
* Installation method used (Homebrew, CocoaPods, building from source, etc)? `mint`
* Paste your configuration file:

```yml
analyzer_rules:
  - unused_import
  - unused_declaration

unused_declaration:
  include_public_and_open: true
```

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)?
  If so, paste their relative paths and respective contents. `No`
* Which Xcode version are you using (check `xcodebuild -version`)? `Xcode 12.5.1`
* Do you have a sample that shows the issue? Yes. Clone https://github.com/acecilia/swift-compilation-database and then do `make go_with_xcodebuild`

**Comment by jpsim** [comment]
I'm aware of this one. It's a SourceKit bug. I don't think I ever reported it to bugs.swift.org though. You're welcome to do that.

**Comment by acecilia** [comment]
Superb, thanks for confirming 🙌 I will try to spend some time and see if there is any workaround to be done

**Comment by jpsim** [comment]
One fairly involved way to fix this would be to use something like SwiftSyntax and index store to build a protocol conformance graph.

**Comment by jpsim** [comment]
It'd be better to file a bug at bugs.swift.org with a less SwiftLint-oriented way this bug manifests itself (e.g. cursor info in Xcode) and fix it at the root.
